### PR TITLE
Refactor PrestashopAutoload system

### PR DESCRIPTION
**Pull request by FabienPapet**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR aims to use only the class index file, other indexes are not useful and don't need to be dumped. This also fixes autoloading namespaced classes without requiring a whole file which include performance as well.
| Type?             | improvement
| Category?         | IN
| BC breaks?        |  no
| Deprecations?     | yes 
| How to test?      | Application should be running without issue, try install/remove/ update modules, overrides system should be working.

## Why refactor the autoload ? 

The current approach has some problems that can't easily be fixed.
- The autoload is *constant dependant* which means if the constants are not defined, the autoload won't work.
- The autoload contains some bugs, for example the override classes are not marked as `override: true` in the `class_index.php` file.
- The core classes folders are hard-coded.
- Many static functions that don't need to be static.
- There are very few tests
- The API is unclear, there is a public function `dumpFile()` that is a copy of the Symfony `FileSystem::dumpFile()` function.
- Autoload is slow: When you require a class from the `Prestashop\Prestashop\Adapter\Entity` namespace, the whole core classes are loaded, even if they are not required.
- Located in the old legacy classes.

## Why create a separated package ?

I've chosen to create a separate package so it can be reused in other future projects. Also, testing such a feature is complicated because you need to update the testsuite just for the autoload as it can create and load new classes. 

## Design choices.

For now, I have split the existing Autoloader into 3 files.

- The first one is the `LegacyClassLoader` which parses the core classes and controllers to build the autoload index. The new class has now configurable directories, removed constants and is fully tested. The class parser is still working as the legacy system.
- The second one is the `Autoloader`. Which loads classes from the index built by the `LegacyClassLoader`.
- The third one is the `PrestashopAutoload` which is the "modernized" autoload.

There is no support for `class_stub.php` and `namespace_class_stubs.php` because they are no more needed in this new version.

## Performance Improvements

These need to be measured with a real tool. But for what i've seen so far, is that the new autoload reduce loading time in admin pages by ~200ms.

## Current problems & next steps. 

- It is difficult to test autoloading classes because in the phpunit bootstrap file uses that same autoload. They may need to be tested separatly.
- PHPStan is based on the old autoload, switching autoload breaks phpstan checks.
- Replace all usages.

### CODE REVIEW

File: src/Core/Autoload/LegacyClassLoader.php
Comment (lartist) [inline src/Core/Autoload/LegacyClassLoader.php]: str_ends_with is available from php8. If I'm not mistaken Prestashop 8.1 should also be compatible with 7.4  and maybe less?

Comment (FabienPapet) [reply src/Core/Autoload/LegacyClassLoader.php]: up to php 7.2 yes ;) . But there is the polyfill given by Symfony that provides the function compatibility for us.

**Comment by kpodemski** [comment]
Hey @FabienPapet 

A little before & after would be very much appreciated :)

**Comment by FabienPapet** [comment]
Sure !

**Comment by FabienPapet** [comment]
Ping @Progi1984 for PHPStan fixes ;)

**Comment by Progi1984** [comment]
Sorry, I'm on tests UI. Have you check the `.github/workflows/phpstan/autoload.php` file (with the line 14 to may be remove) ?

File: src/Core/Autoload/PrestashopAutoloadSingleton.php
Comment (jolelievre) [inline src/Core/Autoload/PrestashopAutoloadSingleton.php]:
```suggestion
final class PrestashopAutoloadFactory
```

Comment (jolelievre) [reply src/Core/Autoload/PrestashopAutoloadSingleton.php]: Suggestion, although it's not only strictly a factory since it acts as a singleton as well, also I wonder if the create method should systematically create a new instance? Or should it return the static cached variable when it's already set?

Another alternative for naming `PrestashopAutoloadBuilder`

**Comment by FabienPapet** [comment]
@kpodemski 
image

**Comment by FabienPapet** [comment]
I continued to work on this and created a separate package to ease reusability and maintenance. I spent too many time fixing  PHPStan issue rather than working on the autoload. 

So I created a separated package from this PR to use this new autoload allowing to use it in outside of prestashop. It is currently in my Github [FabienPapet/Autoload](https://github.com/FabienPapet/autoload) it also has its own composer package name `fpt/autoload` (for now). I'm planning to create a new `prestashop/autoload` package, once my repository is ready.

Integration PR example for the new repository is for now in https://github.com/PrestaShop/PrestaShop/pull/30108 . Once `prestashop/autoload` is ready, I will update this PR and close the other one.

**Comment by FabienPapet** [comment]
@kpodemski can you please create a `PrestaShop\Autoload` Github repository so I can make a PR from my base repo ?

File: classes/PrestaShopAutoload.php
Comment (jolelievre) [inline classes/PrestaShopAutoload.php:167]:
```suggestion
            $legacyClass = substr($className, str_len('PrestaShop\PrestaShop\Adapter\Entity'));
```

Comment (jolelievre) [reply classes/PrestaShopAutoload.php:167]: This would make the code more understandable rather than an arbitrary hard-coded 37
And `'PrestaShop\PrestaShop\Adapter\Entity'` could probably be placed into a private const (not sure it needs to be public)

File: classes/PrestaShopAutoload.php
Comment (jolelievre) [inline classes/PrestaShopAutoload.php:187]: Seems like this method doesn't do the same thing it used to It previously outputted the generated content into a dump file.

Now it only creates the content without exporting it, isn't this a major (breaking) change in this class's behavior? Maybe for some reason it's not needed anymore?

File: classes/PrestaShopAutoload.php
Comment (jolelievre) [inline classes/PrestaShopAutoload.php:70]: I can't comment on the class name but this is a generic comment related with this class. Since you exported your code in an external library if we follow the idea shouldn't this class be removed as well and replaced with the one from the library?
https://github.com/FabienPapet/PrestaShop/blob/a0d7fb0df20149e8c71c0b982e6dbd255f830333/classes/PrestaShopAutoload.php#L34

File: composer.json
Comment (jolelievre) [inline composer.json]: This will, of course, have to be brought back into the PrestaShop organization before this PR can be merged.

**Review by jolelievre — COMMENTED** [review]
Thanks @FabienPapet 

Just some minor comments and questions. The main thing that troubles me though is that we can't review the library's code... I know you put everything in a commit but it's really not convenient for review and it hides most of the complexity of this change.

Do you think you could create a dedicated branch or your repo, empty the main branch and create a pull request? So that we can actually review the whole initial commit on the library?

File: src/Core/Autoload/PrestashopAutoloadSingleton.php
Comment (FabienPapet) [reply src/Core/Autoload/PrestashopAutoloadSingleton.php]: this class does not exist anymore :) i've removed it last week.

**Comment by FabienPapet** [comment]
PR Created for review https://github.com/FabienPapet/autoload/pull/1

File: classes/PrestaShopAutoload.php
Comment (FabienPapet) [reply classes/PrestaShopAutoload.php:70]: It can, but I did not do that because removing code is mainly on major versions.

Comment (jolelievre) [reply classes/PrestaShopAutoload.php:70]: That makes sense! Then maybe we should add a deprecation comment indicating that should be replaced by the class from the library in 9.0 ?

File: src/Core/Autoload/PrestashopAutoloadSingleton.php
Comment (jolelievre) [reply src/Core/Autoload/PrestashopAutoloadSingleton.php]: Yeah sorry, it as an old comment from a previous pending review

File: classes/PrestaShopAutoload.php
Comment (FabienPapet) [reply classes/PrestaShopAutoload.php:167]: LGTM, will fix it once the other repository is ready

File: classes/PrestaShopAutoload.php
Comment (FabienPapet) [reply classes/PrestaShopAutoload.php:187]: The build classIndex is also dumping the index, see LegacyClassLoader.php:77

File: classes/PrestaShopAutoload.php
Comment (FabienPapet) [reply classes/PrestaShopAutoload.php:70]: Deprecate was planned for a future PR, but I can do it if needed

**Comment by FabienPapet** [comment]
hey @PrestaShop/prestashop-core-developers I need your review on this please :)

**Comment by FabienPapet** [comment]
Hey @PrestaShop/prestashop-core-developers , can I have a review on this so it is merged before PrestaShop 8.1, it would be nice to get this merged before code freeze

**Comment by atomiix** [comment]
@FabienPapet looks good to me, can you fix the conflict though?

**Review by nicosomb — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Comment by MhiriFaten** [comment]
Hello @FabienPapet ,

Could you please resolve the conflict? 

image

Thanks.

**Comment by FabienPapet** [comment]
Hey @MhiriFaten,

PR was already rebased 2 days ago, rebasing it will loose the approvals I get,  can you please test it and i'll rebase once everything is ok ? The conflict comes from a dev tool that has been updated.

Thank you !

**Comment by MhiriFaten** [comment]
After other tries to test this PR it can't finish the rebase, it always gives an error.
So we are going to assign another QA on it as long as I have an installation problem and it doesn't seem to work.

Thank you!

**Comment by paulnoelcholot** [comment]
Hello @FabienPapet,

It is not possible for us to test the PR without the rebase, neither manually nor in auto-test. You must rebase the PR.
I'm sorry but we don't have the choice.

**Comment by kpodemski** [comment]
@FabienPapet please rebase, let's approve on Monday, and let's make sure to block any other `composer.lock` PR for a moment :)

**Review by kpodemski — APPROVED** [review]

**Review by mflasquin — APPROVED** [review]

**Comment by FabienPapet** [comment]
ping @PrestaShop/qa-team for tests, it has been rebased

**Review by nicosomb — APPROVED** [review]

**Review by matks — APPROVED** [review]

**Review by MhiriFaten — APPROVED** [review]
Hello @FabienPapet ,

I checked this PR and the application running without issue :tada: 

Checked 

- install modules
- remove modules
- update modules

Overrides system working.
The automated tests are OK.

So it's **QA approved** :heavy_check_mark: 
Thank you.

**Pull request merged**
